File: Dev/unitTest_Engine_cpp_gtest/FamilySystem/AbsoluteParameter.cpp
#include <gtest/gtest.h>
#include <Altseed.h>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class FamilySystem_AbsoluteParameter : public EngineTest
{
public:
	FamilySystem_AbsoluteParameter(bool isOpenGLMode)
		: EngineTest(asd::ToAString("AbsoluteParameter"), isOpenGLMode, 60)
	{
	}

protected:
	void OnStart()
	{
		auto shape = make_shared<CircleShape>();
		shape->SetOuterDiameter(100);

		auto parent = make_shared<TextureObject2D>();
		parent->SetPosition(Vector2DF(40, 40));
		parent->SetColor(Color(128, 255, 255, 255));
		parent->SetDrawingPriority(2);
		parent->SetIsUpdated(false);
		parent->SetIsDrawn(false);

		auto child = make_shared<TextureObject2D>();
		child->SetPosition(Vector2DF(100, 0));
		child->SetColor(Color(255, 255, 128, 255));
		child->SetDrawingPriority(1);
		child->SetIsUpdated(true);
		child->SetIsDrawn(true);

		parent->AddDrawnChild(child,
			ChildManagementMode::IsDrawn | ChildManagementMode::IsUpdated,
			ChildTransformingMode::All,
			ChildDrawingMode::Color | ChildDrawingMode::DrawingPriority);

		Engine::AddObject2D(parent);
		Engine::AddObject2D(child);

		ASSERT_TRUE(child->GetGlobalPosition() == Vector2DF(140, 40));
		ASSERT_EQ(child->GetAbsoluteBeingUpdated(), false);
		ASSERT_EQ(child->GetAbsoluteBeingDrawn(), false);
		ASSERT_TRUE(child->GetAbsoluteColor() == Color(128, 255, 128, 255));
		ASSERT_EQ(child->GetAbsoluteDrawingPriority(), 3);
	}

	void OnUpdating()
	{
	}

	void OnUpdated()
	{
	}

	void OnFinish()
	{
	}
};

ENGINE_TEST(FamilySystem, AbsoluteParameter)

File: Dev/unitTest_Engine_cs/FamilySystem/AbsoluteParameter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.FamilySystem
{
    class AbsoluteParameter : EngineTest
    {
        public AbsoluteParameter() : base(60)
        {
        }

	    private TextureObject2D child;
	    private TextureObject2D child2;

        protected override void OnStart()
        {
            var parent = new TextureObject2D()
            {
                Position = new Vector2DF(40, 40),
                Color = new Color(128, 255, 255, 255),
                DrawingPriority = 2,
                IsUpdated = false,
                IsDrawn = false,
            };
            child = new TextureObject2D()
            {
                Position = new Vector2DF(100, 0),
                Color = new Color(255, 255, 128, 255),
                DrawingPriority = 1,
                IsUpdated = true,
                IsDrawn = true,
            };
            child2 = new TextureObject2D()
            {
                Position = new Vector2DF(0, 100),
                Color = new Color(255, 255, 128, 255),
                DrawingPriority = 1,
                IsUpdated = true,
                IsDrawn = true,
            };
            parent.AddDrawnChild(child,
                ChildManagementMode.IsDrawn | ChildManagementMode.IsUpdated,
                ChildTransformingMode.All,
                ChildDrawingMode.Color | ChildDrawingMode.DrawingPriority);
            parent.AddDrawnChild(child2,
                ChildManagementMode.Nothing,
                ChildTransformingMode.All,
                ChildDrawingMode.Nothing);

            Engine.AddObject2D(parent);
            Engine.AddObject2D(child);
            Engine.AddObject2D(child2);
        }

	    protected override void OnUpdated()
	    {
		    if (Time == 1)
			{
				Assert.AreEqual(child.GetGlobalPosition(), new Vector2DF(140, 40));
				Assert.AreEqual(child.AbsoluteBeingUpdated, false);
				Assert.AreEqual(child.AbsoluteBeingDrawn, false);
				Assert.AreEqual(child.AbsoluteColor, new Color(128, 255, 128, 255));
				Assert.AreEqual(child.AbsoluteDrawingPriority, 3);

				Assert.AreEqual(child2.GetGlobalPosition(), new Vector2DF(40, 140));
				Assert.AreEqual(child2.AbsoluteBeingUpdated, true);
				Assert.AreEqual(child2.AbsoluteBeingDrawn, true);
				Assert.AreEqual(child2.AbsoluteColor, new Color(255, 255, 128, 255));
				Assert.AreEqual(child2.AbsoluteDrawingPriority, 1);
			}
	    }
    }
}
